model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php

```diff
@@ -0,0 +1,131 @@
+
+declare(strict_types=1);
+
+namespace oat\taoQtiItem\model\compile\QtiAssetCompiler;
+
+use InvalidArgumentException;
+use oat\oatbox\config\ConfigurationService;
+use oat\oatbox\filesystem\Directory;
+use oat\taoItems\model\media\ItemMediaResolver;
+use oat\taoQtiItem\model\compile\QtiItemCompilerAssetBlacklist;
+use oat\taoQtiItem\model\pack\QtiAssetPacker\PackedAsset;
+use oat\taoQtiItem\model\qti\AssetParser;
+use oat\taoQtiItem\model\qti\Item;
+use oat\taoQtiItem\model\qti\XIncludeLoader;
+
+class QtiItemAssetCompiler extends ConfigurationService
+{
+    public function extractAndCopyAssetFiles(Item $qtiItem, Directory $publicDirectory, ItemMediaResolver $resolver): array
+    {
+        $assetParser = new AssetParser($qtiItem, $publicDirectory);
+        $assetParser->setGetSharedLibraries(false);
+        $assetParser->setGetXinclude(true);
+
+        $xincludeLoader = new XIncludeLoader($qtiItem, $resolver);
+        $xincludeLoader->load();
+
+        $replacementList = $packedAssets = [];
+
+        foreach ($assetParser->extract() as $type => $assets) {
+
+            foreach ($assets as $key => $assetUrl) {
+
+                if ($this->isBlacklisted($assetUrl)) {
+                    continue;
+                }
+
+                $packedAsset = $this->resolve($resolver, $assetUrl, $type);
+
+                $replacement = $this->getReplacementName($packedAsset->getLink(), $replacementList);
+                $packedAsset->setReplacedBy($replacement);
+
+                $replacementList[$assetUrl] = $replacement;
+
+                if ($type != 'xinclude') {
+                    $this->copyAssetFileToPublicDirectory($publicDirectory, $packedAsset);
+                }
+
+                $packedAssets[$assetUrl] = $packedAsset;
+            }
+        }
+
+        return $packedAssets;
+    }
+
+    private function isBlacklisted(string $assetUrl): bool
+    {
+        return $this->getServiceLocator()
+            ->get(QtiItemCompilerAssetBlacklist::SERVICE_ID)
+            ->isBlacklisted($assetUrl);
+    }
+
+    private function resolve(ItemMediaResolver $resolver, string $assetUrl, string $type): PackedAsset
+    {
+        $mediaAsset = $resolver->resolve($assetUrl);
+        $mediaSource = $mediaAsset->getMediaSource();
+
+        $fileInfo = $mediaSource->getFileInfo($mediaAsset->getMediaIdentifier());
+        if (isset($fileInfo['link'])) {
+            $link = $fileInfo['link'];
+        } elseif(isset($fileInfo['filePath'])) {
+            $link = $fileInfo['filePath'];
+        } else {
+            throw new InvalidArgumentException(sprintf('Item asset %s cannot be resolved.', $assetUrl));
+        }
+
+        return new PackedAsset($type, $mediaAsset, $link);
+    }
+
+    private function getReplacementName(string $link, array $replacementList): string
+    {
+        $basename = basename($link);
+
+        $replacement = $basename;
+
+        $count = 0;
+        while (in_array($replacement, $replacementList)) {
```

Comment:
Code climate is broken because of the complexity of this method. Consider to improve readability.

Reply:
I have isolated this part of code to dedicated method because it was duplicated before